### DISCUSSION TITLE
Project Report - Torch compile and PyTorch 2.2.0 | GSoC 2024

### DISCUSSION THREAD

**Discussion by rida151** [deepchem/deepchem#5156] [Show and tell]
_Note: This post is originally written by @gauthamk02 and has been reposted from DeepChem Forum to GitHub discussions._

I was selected as the contributor for the project “Torch Compile and PyTorch 2.2.0” for Google Summer of Code 2024 under DeepChem. The goal of this project — mentored by Arun Pa and Dr. Bharath Ramsundar — was to integrate the torch.compile() into the DeepChem library. By leveraging this model optimization feature, DeepChem users can now benefit from significantly improved performance during both the training and inference stages of their deep learning models.

PyTorch introduced the torch.compile() function in PyTorch 2.0 to allow faster training and inference of the models. The function works by compiling PyTorch code into optimised kernels using a JIT(Just in Time) compiler. Different models show varying levels of improvement in run times depending on their architecture and batch size when compiled. Compared to existing methods like [TorchScript](https://pytorch.org/docs/stable/jit.html) or [FX tracing](https://pytorch.org/docs/stable/fx.html), compile() also offers advantages such as the ability to handle arbitrary Python code and conditional graph-breaking flow of the inputs to the models. This allows compile() to work with minimal or no code modification to the model. For more information on the compilation process, refer to [PyTorch2.0 Introductory Blog](https://pytorch.org/get-started/pytorch-2.0/) — which does a deep dive into the compilation process, technical decisions and future features for the compile function.

As part of this project, I integrated torch.compile() into Deepchem’s TorchModel class allowing all PyTorch models in DeepChem to be compiled. Additionally, the tutorial [“Compiling Deepchem Torch Models”](https://deepchem.io/tutorials/compiling-deepchem-torch-models/) was also added to DeepChem showing how to compile the models and benchmark them. The weekly progress on the project was tracked on [this](https://forum.deepchem.io/t/torch-compile-and-pytorch-2-2-0-gsoc-2024/1356) forum thread. You can find the official GSoC project page of this project [here](https://summerofcode.withgoogle.com/programs/2024/projects/Xwr0Z1tZ).

### Connect With Me:
Twitter: https://x.com/gaushn_
GitHub: [gauthamk02 (Gautham Krishnan) · GitHub](https://github.com/gauthamk02)
Website: https://gauthamk02.github.io/

### Pull Requests
Adding support for torch.compile() and its tutorial to DeepChem involved several PRs which are:

1. Initial implementation of the compile function and corresponding tests - [link](https://github.com/deepchem/deepchem/pull/3987)
This PR added compile() function to the TorchModel class and also added the corresponding unit tests. The implementation of the function in this PR only allowed inductor backend and default, max-autotune-no-cudagraphs options for mode parameter to be used as these options don’t need any additional dependencies.
2. Adding remaining backends and modes to the function and simplifiying it - [link](https://github.com/deepchem/deepchem/pull/4006)
All the restrictions on the backends and modes on the function was removed in this PR. The users will get an error message if a configuration is used that requires additional dependencies.
3. Add the tutorial for compiling Deepchem models - [link](https://github.com/deepchem/deepchem/pull/4031)
The tutorial for compiling models was added to the DeepChem Tutorials. The tutorial covers:
- Steps to compile DeepChem models.
- The basic theory behind model compilation.
- Methods for comparing the performance of compiled and uncompiled models for training and inference.
- Key considerations for optimizing model compilation.
4. Modify the tutorial to include more contextual information - [link](https://github.com/deepchem/deepchem/pull/4062)
The tutorial was modified to include more contextual information about the need of optimization.
### Benchmarking Performance Improvements
The below results were taken from the tutorial, [“Compiling Deepchem Torch Models”](https://deepchem.io/tutorials/compiling-deepchem-torch-models/), which was added as part of this GSoC project to the Deepchem Tutorials. Refer to the tutorial for the complete code for benchmarking, training and obtaining these results.

To check the performance improvement after compilation, [DMPNN](https://deepchem.readthedocs.io/en/latest/api_reference/models.html#dmpnnmodel) model along with [Freesolv](https://deepchem.readthedocs.io/en/latest/api_reference/moleculenet.html#freesolv-dataset) dataset was used in the tutorial. The uncompiled execution is called Eager Mode and the compiled execution is called Compiled Mode in following results.

Kernel compilations in compiled models cause a significant performance overhead in the starting few steps. To account for this, median values are used as the performance metric when calculating speedup. After the kernels are fully compiled, the time taken by the compiled model per batch stabilises below the uncompiled model. This is because the compilation is done JIT when the model is first run an
<img width="821" height="248" alt="Screenshot 2026-09-25 231337" src="https://github.com/user-attachments/assets/a9a18683-de1f-4877-b2b8-30605931869e" />
d the optimized kernels are generated after a few passes.

The benchmarking results show a clear speedup in the median time per batch for both training and inference after model compilation, demonstrating the effectiveness of torch.compile() in optimizing performance.

### Other Works
During the rest of the program, I have worked on other tasks in DeepChem such as profiling the fit() function and solving other open issues. The profiling was done using [cProfile](https://docs.python.org/3/library/profile.html) and visualized using [SnakeViz](https://jiffyclub.github.io/snakeviz/). However, I am not including the details of that in this report as it was not part of the original GSoC project. You can refer to this [document](https://docs.google.com/document/d/16h1ypzS2roEzT0ZBmx2y0kau4XMaprcXxyy1UePP_5A/edit?usp=sharing) for the results, inferences and code for profiling.
